Paper: pyAudioProcessing: Audio Processing, Feature Extraction and Building Machine Learning Models (#682)
scipy-conference/scipy_proceedings: papers/jyotika_singh/jsingh.rst

research in the audio and speech processing domain. On the contrary, Python remains
the language of choice for a vast majority of Machine Learning research and
functionality. This library contains features built in Python that were originally
published in MATLAB following research inventions. This library allows the user to
extract features such as Mel Frequency Cepstral Coefficients (MFCC), Gammatone
Frequency Cepstral Coefficients (GFCC), spectral features, chroma features and other
beat based and cepstrum-based features from audio to use with one’s own classification
backend or popular scikit-learn classifiers that have been built into pyAudioProcessing.
This software aims to provide machine learning engineers, data scientists, researchers,
and students with a set of baseline models to classify audio, visualize audio
spectrograms and time domain signals, the ability to use this library to build features
on custom training data, automatically train on a scikit-learn classifier and perform
hyper-parameter tuning using this library, export the built features for integration with
any machine learning backend and classify audio files. The library is available at

> https://github.com/jsingh811/pyAudioProcessing and is under GPL-3.0 license.

Reviewer: I believe you want to indent this whole block to be part of the abstract.

Furthermore, audio processing and feature extraction research is popular in 20
MATLAB. There are comparatively fewer resources for audio processing and 21
classification in Python. This tool contains implementation of popular and different 22
audio feature extraction that can be use in combination with most scikit-learn classifiers. 23
Audio data can be trained, tested and classified using pyAudioProcessing [2]. The 24
output consists of cross validation scores and results of testing on custom audio files. 25
PyAudioProcessing is an end-to-end solution for converting between audio file 26
formats, visualizing time and frequency domain representations, silence and low-activity 27
segments removal from audio, building features from raw audio samples and training a 28
machine learning model that can then be used to classify unseen raw audio samples. 29
This library allows the user to extract features such as MFCC, GFCC, spectral features, 30
chroma features and other beat based and cepstrum based features from audio to use 31
with one’s own classification backend or popular scikit-learn classifiers that have been 32

> built into pyAudioProcessing.

Reviewer: this whole paragraph have trailing numbers, issue with copy past ?

:author: Jyotika Singh
:email: singhjyotika811@gmail.com
:institution: Placemakr
:orcid: 0000-0002-5442-3004

:bibliography: mybib

--------------------------------------------------------------------------------------------
pyAudioProcessing: Audio Processing, Feature Extraction and Building Machine Learning Models
--------------------------------------------------------------------------------------------

.. class:: abstract

    PyAudioProcessing is a Python based library for processing audio data, constructing
    and extracting numerical features from audio, building and testing machine learning
    models, and classifying data with existing pre-trained audio classification models or
    custom user-built models. MATLAB is a popular language of choice for a vast amount of
    research in the audio and speech processing domain. On the contrary, Python remains

>     the language of choice for a vast majority of Machine Learning research and

Reviewer 2: ```suggestion
    the language of choice for a vast majority of machine learning research and
```

    functionality. This library contains features built in Python that were originally

>     published in MATLAB following research inventions. PyAudioProcessing allows the user to

Reviewer 2: What does "following research inventions" here mean? What do you want this sentence to convey to the reader?

Reviewer 2: Thanks for looking into this. I see that the sentence was updated to "This library contains features built in Python that were originally published in MATLAB as a part of the original research." I'm still unsure of what this means so unfortunately cannot give a suggestion for how to rephrase it. Could you please help me understand the intent of the sentence? A few questions that come to mind are:
- What "original research" is being referred to? Are there any references to the research that could be included to clarify?
- Was the research done by programmers while they were creating the software Matlab? Or did the programmers read some research papers, then implement the concepts in Matlab? Or something else?
- Why is it important to mention "as part of the original research"? What do you want the reader to take from the phrase or the sentence as a whole?

Author: Thank you for pointing this out. Reading it again, I see your point and agree that the line "as part of original research" does not convey any different information in this paragraph. My original intent was to communicate `software created as a part of performing research or after reading other research papers`. However, I see it isn't sitting well here. As a correction attempt, I've removed the line `as part of original research`.

Reviewer 2: Thanks, @jsingh811. I think the abstract reads smoothly now.

    compute various features from audio files including Gammatone Frequency Cepstral
    Coefficients (GFCC), Mel Frequency Cepstral Coefficients (MFCC), spectral features,

>     chroma features, and others such as beat based and cepstrum-based features from audio.

Reviewer 2: ```suggestion
    chroma features, and others such as beat-based and cepstrum-based features from audio.
```

    One can use these features along with one’s own classification backend or any of the
    popular scikit-learn classifiers that have been integrated into pyAudioProcessing.
    Cleaning functions to strip unwanted portions from the audio are another offering of the library.

>     It further contains integrations with other audio functionalities such as frequency and time series

Reviewer 2: ```suggestion
    It further contains integrations with other audio functionalities such as frequency and time-series
```

    visualizations and audio format conversions. This software aims to provide
    machine learning engineers, data scientists, researchers, and students with a set of baseline models
    to classify audio. The library is available at https://github.com/jsingh811/pyAudioProcessing
    and is under GPL-3.0 license.

.. class:: keywords

   pyAudioProcessing, audio processing, audio data, audio classification, audio feature extraction,
   gfcc, mfcc, spectral features, spectrogram, chroma

Introduction
============

The motivation behind this software is understanding the popularity of Python for

> Machine Learning and presenting solutions for computing complex audio features using

Reviewer 2: Was one of the motivations in creating `pyAudioProcessing` really to understand the **_popularity_** of "Python for Machine Learning"? Could you please rephrase this sentence to clarify the intended meaning?

Reviewer 2: Thanks for updating this!

Python. This not only implies the need for resources to guide solutions for audio
processing, but also signifies the need for Python guides and implementations to solve
audio and speech cleaning, transformation, and classification tasks.

Different data processing techniques work well for different types of data. For

> example, word vector formations work great for text data :cite:`nlp`. However, passing

Reviewer 2: "work great for text data" -> why or in what way? 

Perhaps you can add a short description of what is a "word vector formation" so that the reader, who may not be familiar with word embeddings, can better understand what it would mean to pass "numbers data, an audio signal or an image through word vector formation". I think example(s) of "different data processing techniques" working well for "different types of data" could be developed more to support the main point of the paragraph.

numbers data, an audio signal or an image through word vector formation is not likely
to bring back any meaningful numerical representation that can be used to train
machine learning models. Different data types correlate with feature formation
techniques specific to their domain rather than a one-size-fits-all. These methods for
audio signals are very specific to audio and speech signal processing, which is a domain
of digital signal processing. Digital signal processing is a field of its own and is not

> feasible to master in an ad-hoc fashion. This calls for the need to have popular and

Reviewer 2: "popular" gets used quite a few times throughout the text - is there a more precise term or synonym that could be used here or in other places?

useful processes for audio signals to be in a ready-to-use state by users.

There are two popular approaches for feature building in audio classification tasks.

1. Computing spectrograms from audios as images and using an image classification
pipeline for the remainder.

2. Computing features from audio files directly as numerical vectors and applying

> it to a classification backend.

Reviewer 2: ```suggestion
them to a classification backend.
```

> PyAudioProcessing includes the capability of computing spectrograms, but mainly

Reviewer 2: ```suggestion
PyAudioProcessing includes the capability of computing spectrograms, but
```

> focusses most functionalities around the latter for building audio models. This tool contains implementation

Reviewer 2: ```suggestion
focuses most functionalities around the latter for building audio models. This tool contains implementations
```

> of popular and different audio feature extraction along with integration with popular scikit-learn classifiers.

Reviewer 2: ```suggestion
of various widely used audio feature extraction techniques, and integrates with popular scikit-learn classifiers.
```

How did you choose which scikit-learn classifiers to integrate with? It might be worth listing them somewhere in the paper. Can a user extend the code to integrate with others as needed?

Reviewer 2: Thanks for rephrasing the sentence. I'm not sure if you also saw the questions in my comment above?

Author: Thanks for the questions @mepa. Apologies, unsure how I missed this the first time. 
I just added the list of the classifiers currently supported. The choice was based on starting with the commonly used classic ML model. The list is not constant and is expected to be grow with new code updates to the project.
The user can't extend the code to integrate with other models for hyperparameter tuning in the current version of the library. The way to use any other models for now would be extracting the features and passing that to the classifier you want to use.

Reviewer 2: No problem, @jsingh811, and thanks for adding the list of currently supported classifiers. I appreciate your explanation of how one can, for now, use other models, too.

Audio data can be cleaned, trained, tested, and classified using pyAudioProcessing :cite:`pAP`.

Some other useful libraries for the domain of audio processing include librosa

> :cite:`mcfee2015librosa`, spafe, essentia :cite:`essentia`,

Reviewer 2: The citation for spafe is shown here: https://github.com/SuperKogito/spafe/blob/master/docs/source/index.rst

pyAudioAnalysis :cite:`giannakopoulos2015pyaudioanalysis`, and paid services from service
providers such as Google [#]_.

.. [#] https://developers.google.com/learn/pathways/get-started-audio-classification

The use of pyAudioProcessing in the community inspires the need and growth of this software.
It is referenced in a text book titled `Artificial Intelligence with Python Cookbook` published by
Packt Publishing in October 2020 :cite:`packt`. Additionally, pyAudioProcessing is a part of specific
admissions requirement for a funded PhD project at University of Portsmouth [#]_.
It is further referenced in this thesis paper titled "Master Thesis AI Methodologies for Processing

> Acoustic Signals AI Usage for Processing Acoustic Signals" :cite:`phdthesis`, in recent researches in audio

Reviewer 2: ```suggestion
Acoustic Signals AI Usage for Processing Acoustic Signals" :cite:`phdthesis`, in recent research of audio
```

processing for assessing attention levels in Attention Deficit Hyperactivity Disorder (ADHD)

> students :cite:`paper`, and more.

Reviewer 2: Along with these specific examples, perhaps download statistics could also be used to motivate the need and growth of pyAudioProcessing: https://pypistats.org/packages/pyaudioprocessing. What kind of growth is needed for this package and how can open-source contributors help?

Reviewer 2: Thanks for adding that context.

.. [#] https://www.port.ac.uk/study/postgraduate-research/research-degrees/phd/explore-our-projects/detection-of-emotional-states-from-speech-and-text

Core Functionalities
====================

PyAudioProcessing aims to provide an end-to-end processing solution for converting between audio file

> formats, visualizing time and frequency domain representations, cleaning with silence and low-activity

Reviewer 2: Which formats, and is the conversion one-way (e.g., everything to .wav) or two-way (e.g., .mp3 <--> .wav)?

segments removal from audio, building features from raw audio samples, and training a

> machine learning model that can then be used to classify unseen raw audio samples.

Reviewer 2: "classify unseen raw audio samples" --> I think it might be helpful to give examples at this point of classification categories such as music/speech, or music genre.

This library allows the user to extract features such as Mel Frequency Cepstral Coefficients (MFCC) :cite:`6921394`,
Gammatone Frequency Cepstral Coefficients (GFCC) :cite:`inbook`, spectral features,

> chroma features and other beat based and cepstrum based features from audio to use

Reviewer 2: ```suggestion
chroma features and other beat-based and cepstrum-based features from audio to use
```

> with one’s own classification backend or popular scikit-learn classifiers that have been

Reviewer 2: Remove "popular" here?

> built into pyAudioProcessing. The classifier implementation examples that are a part of this software aim to give the users

Reviewer 2: ```suggestion
built into pyAudioProcessing. The classifier implementation examples that are a part of this software aim to give users
```

a sample solution to audio classification problems and help build the foundation to
tackle new and unseen problems.

PyAudioProcessing provides seven core functionalities comprising different stages of audio signal processing.

1. Converting audio files to .wav  format to give the users the ability to work with different types of audio
to increase compatibility with code and processes that work best with .wav audio type.

2. Audio visualization in time-series and frequency representation, including
spectrograms.

3. Segmenting and removing low-activity segments from audio files for removing unwanted audio segments that are
less likely to represent meaningful information.

4. Building numerical features from audio that can be used to train machine learning models. The set of features

> supported evolve with time as research informs new and improved algorithms.

Reviewer 2: ```suggestion
supported evolves with time as research informs new and improved algorithms.
```

5. Ability to export the features built with this library to use with any custom machine learning backend of
the user's choosing.

6. Capability that allows users to train scikit-learn classifiers using features of their choosing directly

> from raw data. This library runs

Reviewer 2: ```suggestion
from raw data. This library
```

Also, which library? scikit-learn or pyAudioProcessing?

>   a). automatic hyper-parameter tuning

Reviewer 2: ```suggestion
  a). runs automatic hyper-parameter tuning
```

>   b). returns to the user the training model metrics along with cross-validation confusion matrix for model evaluation

Reviewer 2: "cross-validation confusion matrix" -> define this term?

>   c). allows the users to test the created classifier with the same features used for training

Reviewer 2: ```suggestion
  c). allows users to test the created classifier with the same features used for training
```

7. Includes pre-trained models to provide users with baseline audio
classifiers.

Methods and Results
===================

Pre-trained models
------------------

> This software offers pre-trained audio classification models for the Python community to aid in quick baseline establishment.

Reviewer 2: ```suggestion
PyAudioProcessing offers pre-trained audio classification models for the Python community to aid in quick baseline establishment.
```

This is an evolving feature as new datasets and classification problems gain prominence in the field.

Some of the pre-trained models include the following.

1. Audio type classifier to determine speech versus music:

> Trained a Support Vector Machine (SVM) classifier for classifying audio into two possible classes - music,

Reviewer 2: Which data was the model trained on?

Reviewer 2: Thanks for adding that!

speech. This classifier was trained using mel frequency cepstral coefficients (MFCC), spectral, and chroma features.

> Confusion matrix produces scores as shown in :ref:`speech-music`.

Reviewer 2: ```suggestion
The confusion matrix produces percentages as shown in :ref:`speech-music`.
```

I don't think I'd refer to confusion matrix values (or percentages, in this case) as scores. This suggestion could also be applied to the other pre-trained model descriptions below.

.. table:: Confusion matrix for audio type (speech vs music) classification pre-trained model. :label:`speech-music`

     +-----------+------------+------------+
     |           | **Music**  | **Speech** |
     +-----------+------------+------------+
     | **Music** | **48.80**  | 1.20       |
     +-----------+------------+------------+
     | **Speech**| 0.60       | **49.40**  |

>      +-----------+------------+------------+

Reviewer 2: How many total samples were classified? Since this and the other confusion matrices show percentages instead of counts, I'd mention it in the caption of the figure and/or add `%` after the numbers in all cases.

2. Audio type classifier to determine speech versus music versus bird sounds:

> Trained Support Vector Machine (SVM) classifier that classifying audio into three possible classes -

Reviewer 2: ```suggestion
Trained Support Vector Machine (SVM) classifier for classifying audio into three possible classes -
```

> music, speech, birds. This classifier was trained using mel frequency cepstral coefficients (MFCC), spectral and

Reviewer 2: ```suggestion
music, speech, birds. This classifier was trained using mel frequency cepstral coefficients (MFCC), spectral features, and
```

chroma features. Confusion matrix has scores as shown in :ref:`speech-music-birds`.

.. table:: Confusion matrix for audio type (speech vs music vs bird sound) classification pre-trained model. :label:`speech-music-birds`

     +-----------+------------+------------+------------+
     |           | **Music**  | **Speech** | **Birds**  |
     +-----------+------------+------------+------------+
     | **Music** | **31.53**  | 0.73       | 1.07       |
     +-----------+------------+------------+------------+
     | **Speech**| 1.00       | **32.33**  | 0.00       |
     +-----------+------------+------------+------------+
     | **Birds** | 0.00       | 0.00       | **33.33**  |
     +-----------+------------+------------+------------+

3. Music genre classifier using the GTZAN :cite:`tzanetakis:2001`:
Trained on SVM classifier using gammatone frequency cepstral coefficients (GFCC),

> mel frequency cepstral coefficients (MFCC), spectral and chroma features to classify

Reviewer 2: ```suggestion
mel frequency cepstral coefficients (MFCC), spectral features, and chroma features to classify
```

spectral features and chroma features are two separate features, is that right? or is it "spectral and chroma" features?

    +----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+
    | **rock** | 0.14     | 0.90     | 1.10     | 0.80     | 0.35     | 0.29     | **5.31** | 0.01     | 1.09     | 0.01     |
    +----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+
    | **hip**  | 0.71     | 0.14     | 0.56     | 0.18     | 1.96     | 0.00     | 0.19     | **6.10** | 0.03     | 0.14     |
    +----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+
    | **cou**  | 0.25     | 0.15     | 0.84     | 0.64     | 0.08     | 0.10     | 1.87     | 0.00     | **5.84** | 0.24     |
    +----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+
    | **jazz** | 0.04     | 0.01     | 0.13     | 0.41     | 0.00     | 0.76     | 0.31     | 0.00     | 0.53     | **7.81** |
    +----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+

> These baseline models aim to present capability of audio feature generation algorithms

Reviewer 2: I think of a "baseline model" as a simple model trained on a dataset to which one compares performance of more complex models trained on the same dataset. Is that the intended meaning here? Or would it be fine to remove the word "baseline" to avoid confusion?

in extracting meaningful numeric patterns from the audio data. One can train their own
classifiers using similar features and different machine learning backend for researching
and exploring improvements.

Audio features
--------------

There are multiple types of features one can extract from audio. Information about
getting started with audio processing is well described in :cite:`opensource`.
pyAudioProcessing allows users to compute GFCC, MFCC, other cepstral features, spectral features,
temporal features, chroma features, and more. Details on how to extract these features
are present in the project documentation on GitHub. Generally, features useful in different audio prediction

> tasks (especially speech) include Linear prediction coefficients and Linear prediction cepstral coefficients (LPCC),

Reviewer 2: ```suggestion
tasks (especially speech) include linear prediction coefficients and linear prediction cepstral coefficients (LPCC),
```
(assuming those don't need to be capitalized)

> Bark frequency cepstral coefficients (BFCC), Power normalized cepstral coefficients (PNCC), and

Reviewer 2: ```suggestion
Bark frequency cepstral coefficients (BFCC), power normalized cepstral coefficients (PNCC), and
```

spectral features like spectral flux, entropy, roll off, centroid, spread, and energy entropy.

While MFCC features find use in most commonly encountered audio processing tasks such as audio type
classification, speech classification, GFCC features have been found to have application in speaker

> identification or speaker diarization. Applications, comparisons and uses can be found

Reviewer 2: What is diarization?

Reviewer 2: Thanks!

below 1 kHz and logarithmic for frequencies above 1 kHz, as shown in Figure :ref:`S1Fig`.
This is motivated by the fact that the human auditory system becomes less frequency-selective as
frequency increases above 1 kHz.

.. figure:: S1_Fig.png
   :scale: 39%
   :figclass: w

   MFCC from audio spectrum. :label:`S1Fig`

The signal is divided into segments and a spectrum is computed.
Passing a spectrum through the Mel filter bank, followed by taking the log magnitude and a
discrete cosine transform (DCT) produces the Mel cepstrum. DCT extracts the signal's main
information and peaks. For this very property, DCT is also widely used in applications such as
JPEG and MPEG compressions. The peaks after DCT contain the gist of the audio information.
Typically, the first 13-20 coefficients extracted from the Mel cepstrum are called the MFCCs.
These hold very useful information about audio and are often used to train machine learning models.
The process of developing these coefficients can be seen in the form of an illustration in Figure :ref:`S1Fig`.
MFCC for a sample speech audio can be seen in Figure :ref:`mfccspeech`.

.. figure:: mfcc_speech.png
   :scale: 39%

>    :figclass: bht

Reviewer 2: Can't see the title, axes labels, or tick labels in figures 1 and 3.

   MFCC from a sample speech audio. :label:`mfccspeech`

**Gammatone Frequency Cepstral Coefficients (GFCC)**
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Another filter inspired by human hearing is the Gammatone filter bank. The
gammatone filterbank shape looks similar to the Mel filter bank, expect the peaks
are smoother than the triangular shape of the mel filters. Gammatone filters
are conceived to be a good approximation to the human auditory filters and are used as a
front-end simulation of the cochlea. Since a human ear is the perfect receiver and distinguisher
of speakers in the presence of noise or no noise, construction of gammatone filters that mimic
auditory filters became desirable. Thus, it has many applications in speech processing because
it aims to replicate how we hear.

.. figure:: S2_Fig.png
   :scale: 34%

>    :figclass: w

Reviewer 2: In most of the figures the font is quite small. Could font or overall figure size be increased, especially figures 8-11? Also, this and some others are missing axis labels.

   GFCC from audio spectrum. :label:`S2Fig`

> GFCCs are formed by passing the spectrum through Gammatone filter bank, followed by

Reviewer 2: ```suggestion
GFCCs are formed by passing the spectrum through a gammatone filter bank, followed by
```

loudness compression and DCT, as seen in Figure :ref:`S2Fig`. The first
(approximately) 22 features are called GFCCs. GFCCs have a number of applications
in speech processing, such as speaker identification. GFCC for a sample
speech audio can be seen in Figure :ref:`gfccspeech`.

.. figure:: gfcc_speech.png
   :scale: 39%
   :figclass: bht

   GFCC from a sample speech audio. :label:`gfccspeech`

**Temporal Features**
^^^^^^^^^^^^^^^^^^^^^

Temporal features from audio are extracted from the signal information in its time domain representations.
Examples include signal energy, entropy, zero crossing rate, etc.
Some sample mean temporal features can be seen in Figure :ref:`temp`.

.. figure:: temporal_speech.png
   :scale: 39%
   :figclass: bht

   Temporal extractions from a sample speech audio. :label:`temp`

**Spectral features**
^^^^^^^^^^^^^^^^^^^^^

Spectral features on the other hand derive information contained in the frequency domain representation of an audio signal.

> The signal can be converted from time domain to frequency domain using Fourier Transform. Useful

Reviewer 2: ```suggestion
The signal can be converted from time domain to frequency domain using the Fourier transform. Useful
```

features from the signal spectrum include fundamental frequency, spectral entropy, spectral spread, spectral flux,
spectral centroid, spectral roll-off, etc.
Some sample mean spectral features can be seen in Figure :ref:`spec`.

.. figure:: spectral_speech.png
   :scale: 39%
   :figclass: bht

   Spectral features from a sample speech audio. :label:`spec`

**Chroma Features**
^^^^^^^^^^^^^^^^^^^

Chroma features are highly popular for music audio data.
In Western music, the term chroma feature or chromagram closely relates to the twelve different pitch classes.
Chroma-based features, which are also referred to as "pitch class profiles", are a powerful tool for analyzing
music whose pitches can be meaningfully categorized (often into twelve categories : A, A#, B, C, C#, D, D#, E, F, F#, G, G#
) and whose tuning approximates to the equal-tempered scale :cite:`chromawiki`.

> A popular characteristic of chroma features is that they capture the harmonic and melodic attributes of audio,

Reviewer 2: "popular" is used twice in this paragraph

> while being robust to changes in timbre and instrumentation.

Reviewer 2: Really well-described!

Some sample mean chroma features can be seen in Figure :ref:`chroma`.

.. figure:: chroma_speech.png
   :scale: 39%
   :figclass: bht

   Chroma features from a sample speech audio. :label:`chroma`

Audio data cleaning
-------------------

> Often times an audio has multiple segments present in the same signal that do not contain anything but

Reviewer 2: ```suggestion
Often times an audio sample has multiple segments present in the same signal that do not contain anything but
```

silence or a slight degree of background noise compared to the rest of the audio.

> For most applications, those low activity segments make up for the relevant information

Reviewer 2: What is meant by "make up for" here? Should it be "make up"?

of the signal.

The audio clip shown in Figure :ref:`S3Fig` is a human saying the word "london" and represents

> the audio plotted in time-domain, with signal amplitude as y-axis and sample number as x-axis. The

Reviewer 2: "time-domain" -> "the time domain"

various calls of animals. A spectrogram can be generated by an optical spectrometer, a bank of band-pass
filters, by Fourier transform or by a wavelet transform. A spectrogram is usually depicted as a heat map,
i.e., as an image with the intensity shown by varying the color or brightness.

.. figure:: S4_Fig.png
   :scale: 30%
   :figclass: bht

   Spectrogram of speech for "london". :label:`S4Fig`

After applying the algorithm for signal alteration to remove irrelevant and low activity audio
segments, the resultant audio's time-series plot looks like Figure :ref:`S5Fig`. The spectrogram
looks like Figure :ref:`S6Fig`. It can be seen that the low activity areas are now missing from the
audio and the resultant audio contains more activity filled regions. This algorithm removes silences
as well as low-activity regions from the audio.

.. figure:: S5_Fig.png
   :scale: 30%
   :figclass: bht

   Time-series representation of cleaned speech for "london". :label:`S5Fig`

.. figure:: S6_Fig.png
  :scale: 30%
  :figclass: bht

  Spectrogram of cleaned speech for "london". :label:`S6Fig`

These visualizations were produced using pyAudioProcessing and can be produced for any audio signal

> using the library.

Reviewer 2: Nice description of the audio features that can be produced with pyAudioProcessing. Looks useful!

Impact of cleaning on feature formations for a classification task

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer 2: Please check the output pdf for the paper. Is this text^^ rendered as expected?

A spoken location name classification problem was considered for this evaluation.
The dataset consisted of 23 samples for training per class and 17 samples for testing per class.

> The total number of classes are 2 - london and boston. This dataset can be found linked in the project

Reviewer 2: ```suggestion
The total number of classes is 2 - london and boston. This dataset can be found linked in the project
```

> readme of pyAudioProcessing. For comparative purposes, the classifier is kept constant

Reviewer 2: Does the dataset have a DOI or other citation?

at SVM, and the parameter C is chosen based on grid search for each experiment based

> on best precision, recall and F1. Results in table :ref:`clean` show the impact of

Reviewer 2: ```suggestion
on best precision, recall and F1 score. Results in table :ref:`clean` show the impact of
```

applying the low-activity region removal using pyAudioProcessing prior to training
the model using MFCC features.

> It can be seen that the accuracies increased when audios were cleaned prior to training the model.

Reviewer 2: ```suggestion
It can be seen that the accuracies increased when audio samples were cleaned prior to training the model.
```

> This is specially useful in cases where silence or low-activity regions in the

Reviewer 2: ```suggestion
This is especially useful in cases where silence or low-activity regions in the
```

audio do not contribute to the predictions and act as noise in the signal.

.. table:: Performance comparison on test data between MFCC feature trained model with and without cleaning. :label:`clean`

   +------------------+------------------+------------------+
   | Features (Tools) | boston acc       | london acc       |
   +==================+==================+==================+
   | **mfcc**         | 0.765            | 0.412            |
   +------------------+------------------+------------------+
   | **clean+mfcc**   | **0.823**        | **0.471**        |
   +------------------+------------------+------------------+

Integrations
------------

**Training, classification, and evaluation**
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The library contains integrations with scikit-learn classifiers for passing audio

> through feature extraction followed by classification directly using the raw audios

Reviewer 2: ```suggestion
through feature extraction followed by classification directly using the raw audio samples
```

I don't think "audios" is a typical way to use the word, so might be worth changing here and elsewhere throughout the text.

as input. Training results include computation of cross-validation results along
with hyperparameter tuning details.

**Audio format conversion**
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Some applications and integrations work best with .wav data format. pyAudioProcessing

> integrates with tools that perform format conversion and presents it as a functionality

Reviewer 2: ```suggestion
integrates with tools that perform format conversion and presents them as a functionality
```

via the library.

**Audio visualization**
^^^^^^^^^^^^^^^^^^^^^^^

Spectrograms are 2-D images representing sequences of spectra with time along one axis,
frequency along the other, and brightness or color representing the strength of a frequency
component at each time frame :cite:`spectro`. Not only can one see whether there is more or less energy at,
for example, 2 Hz vs 10 Hz, but one can also see how energy levels vary over time :cite:`specPNSN`.
Some of the convolutional neural network architectures for images can be applied to audios on top of

> the spectrograms. This is a different route of building audio models is developing spectrograms

Reviewer 2: "is" is used twice in one sentence

followed by image processing.

> Time-series, frequency-domain, and spectrograms (both time and frequency domains)

Reviewer 2: ```suggestion
Time-series, frequency-domain, and spectrogram (both time and frequency domains) visualizations
```

can be retrieved using pyAudioProcessing and its integrations.
See figures :ref:`S5Fig` and :ref:`S4Fig` as examples.

Conclusion
==========

> In this paper pyAudioProcessing, an open-source Python library, is presented that implements and integrates a

Reviewer 2: Consider splitting the sentence in to two at "that"

wide range of audio processing functionalities. Using pyAudioProcessing, one can read and visualize

> audios, clean audio signals by removal of irrelevant content, build and extract complex features such

Reviewer 2: Why "audios" then "audio signals"? Do these refer to different concepts, or should they both be "audio signals"?

as GFCC, MFCC and other spectrum and cepstrum based features, build classification models,
and use pre-built trained baseline models to classify different types of audio. Wrappers along with
command-line usage examples are provided in the software's readme and wiki for giving the user
flexibility of usage. pyAudioProcessing has been used in active research around audio processing
and can be used as the basis for further python-based research efforts.

pyAudioProcessing is updated frequently in order to apply enhancements and new functionalities
with recent research efforts of the digital signal processing and machine learning community.

> Some of the ongoing implementations include LPCC feature additions and integration with deep learning backend.

Reviewer 2: ```suggestion
Some of the ongoing implementations include LPCC feature additions and integration with a deep learning backend.
```
Or should it be "deep learning backends"? Which one(s)?

Acknowledgments
===============

> This project was built with the consideration for contributions towards the open-source community.

Reviewer 2: I'm not sure I understand the intent of this sentence. Is it implying that the project was built to enable open-source contributions? Or that you built it as a contribution toward the open-source community?

It is further referenced in this thesis paper titled "Master Thesis AI Methodologies for Processing
Acoustic Signals AI Usage for Processing Acoustic Signals" :cite:`phdthesis`, in recent research on audio
processing for assessing attention levels in Attention Deficit Hyperactivity Disorder (ADHD)
students :cite:`paper`, and more. There are thus far 16000+ downloads via pip for pyAudioProcessing with 1000+ downloads in the last month :cite:`pepy`. As several different audio features need development, new issues are created on GitHub and contributions to the code by the open-source community are welcome to grow the tool faster.

.. [#] https://www.port.ac.uk/study/postgraduate-research/research-degrees/phd/explore-our-projects/detection-of-emotional-states-from-speech-and-text

Core Functionalities
====================

PyAudioProcessing aims to provide an end-to-end processing solution for converting between audio file
formats, visualizing time and frequency domain representations, cleaning with silence and low-activity
segments removal from audio, building features from raw audio samples, and training a

> machine learning model that can then be used to classify unseen raw audio samples (e.g., into categories such a music, speech, etc.).

Reviewer 2: ```suggestion
machine learning model that can then be used to classify unseen raw audio samples (e.g., into categories such as music, speech, etc.).
```

the user's choosing.

6. Capability that allows users to train scikit-learn classifiers using features of their choosing directly
from raw data. pyAudioProcessing

  a). runs automatic hyper-parameter tuning

  b). returns to the user the training model metrics along with cross-validation confusion matrix (a cross-validation confusion matrix is an evaluation matrix from where we can estimate the performance of the model broken down by each class/category) for model evaluation

  c). allows the user to test the created classifier with the same features used for training

7. Includes pre-trained models to provide users with baseline audio
classifiers.

Methods and Results
===================

Pre-trained models
------------------

pyAudioProcessing offers pre-trained audio classification models for the Python community to aid in quick baseline establishment.
This is an evolving feature as new datasets and classification problems gain prominence in the field.

Some of the pre-trained models include the following.

1. Audio type classifier to determine speech versus music:
Trained a Support Vector Machine (SVM) classifier for classifying audio into two possible classes - music,

> speech. This classifier was trained using Mel Frequency Cepstral Coefficients (MFCC), spectral features, and chroma features. This model was trained on manually curated custom samples for speech and music.

Reviewer 2: Manually curated, or created?

> The per-class evaluation metrics are shown in :ref:`speech-music`.

Reviewer 2: ```suggestion
The per-class evaluation metrics are shown in Table :ref:`speech-music`.
```

.. table:: Per-class evaluation metrics for audio type (speech vs music) classification pre-trained model. :label:`speech-music`

     +-----------+-------------+-------------+-------------+
     | Class     |             | Metric      |             |
     +-----------+-------------+-------------+-------------+
     |           |**Accuracy** |**Precision**| **F1**      |
     +===========+=============+=============+=============+
     | **music** | 97.60%      | 98.79%      | 98.19%      |
     +-----------+-------------+-------------+-------------+
     | **speech**| 98.80%      | 97.63%      | 98.21%      |
     +-----------+-------------+-------------+-------------+

2. Audio type classifier to determine speech versus music versus bird sounds:
Trained Support Vector Machine (SVM) classifier for classifying audio into three possible classes -
music, speech, birds. This classifier was trained using Mel Frequency Cepstral Coefficients (MFCC), spectral features, and

> chroma features. The per-class evaluation metrics are shown in :ref:`speech-music-birds`.

Reviewer 2: ```suggestion
chroma features. The per-class evaluation metrics are shown in Table :ref:`speech-music-birds`.
```

.. table:: Per-class evaluation metrics for audio type (speech vs music vs bird sound) classification pre-trained model. :label:`speech-music-birds`

     +-----------+-------------+-------------+-------------+
     | Class     |             | Metric      |             |
     +-----------+-------------+-------------+-------------+
     |           |**Accuracy** |**Precision**| **F1**      |
     +===========+=============+=============+=============+
     | **music** | 94.60%      | 96.93%      | 95.75%      |
     +-----------+-------------+-------------+-------------+
     | **speech**| 97.00%      | 97.79%      | 97.39%      |
     +-----------+-------------+-------------+-------------+
     | **birds** | 100.00%     | 96.89%      | 98.42%      |
     +-----------+-------------+-------------+-------------+

3. Music genre classifier using the GTZAN :cite:`tzanetakis:2001`:
Trained on SVM classifier using Gammatone Frequency Cepstral Coefficients (GFCC),
Mel Frequency Cepstral Coefficients (MFCC), spectral features, and chroma features to classify
music into 10 genre classes - blues, classical, country, disco, hiphop,

> jazz, metal, pop, reggae, rock. The per-class evaluation metrics are shown in :ref:`music-class-genre`.

Reviewer 2: ```suggestion
jazz, metal, pop, reggae, rock. The per-class evaluation metrics are shown in Table :ref:`music-class-genre`.
```

frequency increases above 1 kHz.

.. figure:: S1_Fig.png
   :scale: 42%
   :figclass: w

   MFCC from audio spectrum. :label:`S1Fig`

The signal is divided into segments and a spectrum is computed.
Passing a spectrum through the Mel filter bank, followed by taking the log magnitude and a
discrete cosine transform (DCT) produces the Mel cepstrum. DCT extracts the signal's main
information and peaks. For this very property, DCT is also widely used in applications such as
JPEG and MPEG compressions. The peaks after DCT contain the gist of the audio information.
Typically, the first 13-20 coefficients extracted from the Mel cepstrum are called the MFCCs.
These hold very useful information about audio and are often used to train machine learning models.
The process of developing these coefficients can be seen in the form of an illustration in Figure :ref:`S1Fig`.
MFCC for a sample speech audio can be seen in Figure :ref:`mfccspeech`.

.. figure:: mfcc_speech.png
  :scale: 25%
  :figclass: bht

  MFCC from a sample speech audio. :label:`mfccspeech`

|

**Gammatone Frequency Cepstral Coefficients (GFCC)**
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
|

> Another filter inspired by human hearing is the Gammatone filter bank. The

Reviewer 2: ```suggestion
Another filter inspired by human hearing is the gammatone filter bank. The
```

> gammatone filterbank shape looks similar to the Mel filter bank, expect the peaks

Reviewer 2: I'm not sure what's the standard usage (capitalized? is filter bank one word or two? ), so please choose as needed. I think it's important only that they're self-consistent throughout the article.
```suggestion
gammatone filter bank shape looks similar to the Mel filter bank, expect the peaks
```

Chroma-based features, which are also referred to as "pitch class profiles", are a powerful tool for analyzing
music whose pitches can be meaningfully categorized (often into twelve categories : A, A#, B, C, C#, D, D#, E, F, F#, G, G#
) and whose tuning approximates to the equal-tempered scale :cite:`chromawiki`.
A prime characteristic of chroma features is that they capture the harmonic and melodic attributes of audio,
while being robust to changes in timbre and instrumentation.
Some sample mean chroma features can be seen in Figure :ref:`chroma`.

.. figure:: chroma_speech.png
   :scale: 25%
   :figclass: bht

   Chroma features from a sample speech audio. :label:`chroma`

Audio data cleaning
-------------------

Often times an audio sample has multiple segments present in the same signal that do not contain anything but
silence or a slight degree of background noise compared to the rest of the audio.
For most applications, those low activity segments make up the irrelevant information
of the signal.

The audio clip shown in Figure :ref:`S3Fig` is a human saying the word "london" and represents

> the audio plotted in the time-domain, with signal amplitude as y-axis and sample number as x-axis. The

Reviewer 2: ```suggestion
the audio plotted in the time domain, with signal amplitude as y-axis and sample number as x-axis. The
```

   Spectrogram of speech for "london". :label:`S4Fig`

After applying the algorithm for signal alteration to remove irrelevant and low activity audio
segments, the resultant audio's time-series plot looks like Figure :ref:`S5Fig`. The spectrogram
looks like Figure :ref:`S6Fig`. It can be seen that the low activity areas are now missing from the
audio and the resultant audio contains more activity filled regions. This algorithm removes silences
as well as low-activity regions from the audio.

.. figure:: S5_Fig.png
   :scale: 45%
   :figclass: bht

   Time-series representation of cleaned speech for "london". :label:`S5Fig`

.. figure:: S6_Fig.png
  :scale: 45%
  :figclass: bht

  Spectrogram of cleaned speech for "london". :label:`S6Fig`

These visualizations were produced using pyAudioProcessing and can be produced for any audio signal
using the library.

|

**Impact of cleaning on feature formations for a classification task**
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
|

A spoken location name classification problem was considered for this evaluation.
The dataset consisted of 23 samples for training per class and 17 samples for testing per class.

> The total number of classes is 2 - london and boston. This dataset was manually curated and can be found linked in the project

Reviewer 2: Manually curated, or created?

at SVM, and the parameter C is chosen based on grid search for each experiment based
on best precision, recall and F1 score. Results in table :ref:`clean` show the impact of
applying the low-activity region removal using pyAudioProcessing prior to training
the model using MFCC features.

It can be seen that the accuracies increased when audio samples were cleaned prior to training the model.
This is especially useful in cases where silence or low-activity regions in the
audio do not contribute to the predictions and act as noise in the signal.

.. table:: Performance comparison on test data between MFCC feature trained model with and without cleaning. :label:`clean`

   +------------------+------------------+------------------+
   | Features         | boston acc       | london acc       |
   +==================+==================+==================+
   | **mfcc**         | 0.765            | 0.412            |
   +------------------+------------------+------------------+
   | **clean+mfcc**   | **0.823**        | **0.471**        |
   +------------------+------------------+------------------+

Integrations

> ------------

Reviewer 2: The formatting of this section in the rendered pdf looks a bit strange so please take a look, but if it's as expected then no problem.

in :cite:`6639061`, :cite:`patent`, and :cite:`patent2`.

pyAudioProcessing library includes computation of these features for audio segments of a single audio,
followed by computing mean and standard deviation of all the signal segments.

|

Mel Frequency Cepstral Coefficients (MFCC)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
|

The mel scale relates perceived frequency, or pitch, of a pure tone to its actual measured
frequency. Humans are much better at discerning small changes in pitch at low frequencies
compared to high frequencies. Incorporating this scale makes our features match more
closely what humans hear. The mel-frequency scale is approximately linear for frequencies
below 1 kHz and logarithmic for frequencies above 1 kHz, as shown in Figure :ref:`S1Fig`.
This is motivated by the fact that the human auditory system becomes less frequency-selective as
frequency increases above 1 kHz.

.. figure:: S1_Fig.png
   :scale: 42%
   :figclass: w

   MFCC from audio spectrum. :label:`S1Fig`

The signal is divided into segments and a spectrum is computed.

> Passing a spectrum through the Mel filter bank, followed by taking the log magnitude and a

Reviewer 2: ```suggestion
Passing a spectrum through the mel filter bank, followed by taking the log magnitude and a
```

   | Features         | boston acc       | london acc       |
   +==================+==================+==================+
   | **mfcc**         | 0.765            | 0.412            |
   +------------------+------------------+------------------+
   | **clean+mfcc**   | **0.823**        | **0.471**        |
   +------------------+------------------+------------------+

**Integrations**
----------------

pyAudioProcessing integrates with third-party tools such as scikit-learn, matplotlib, and pydub to offer additional functionalities.

|

Training, classification, and evaluation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
|

The library contains integrations with scikit-learn classifiers for passing audio
through feature extraction followed by classification directly using the raw audio samples
as input. Training results include computation of cross-validation results along
with hyperparameter tuning details.

|

Audio format conversion
^^^^^^^^^^^^^^^^^^^^^^^
|

Some applications and integrations work best with .wav data format. pyAudioProcessing
integrates with tools that perform format conversion and presents them as a functionality
via the library.

|

Audio visualization

> ^^^^^^^^^^^^^^^^^^^

Reviewer 2: Please see the built pdf at https://procbuild.scipy.org/download/jsingh811-2022-jsingh-paper and consider adding or removing text somewhere to force this text to be placed above the paragraph it heads. No worries if it doesn't work out, but just pointing it out in case you'd like to fix it.

Author: Done👍
